timemachine/cpp/src/barostat.cu

```diff
@@ -172,41 +202,115 @@ void __global__ find_group_centroids(
+    const double new_volume = volume+delta_volume;
+    d_volume_delta[0] = delta_volume;
+    d_length_scale[0] = cbrt(new_volume/volume);
 }
 
-void MonteCarloBarostat::inplace_move(
-    double *d_x,
-    double *d_box,
-    const double lambda
-    ) {
+void __global__ k_decide_move(
+    const int N,
+    const int num_molecules,
+    const double kt,
+    const double pressure,
+    const double * __restrict__ rand, // [2] Use second value
+    double * __restrict__ d_volume_delta,
+    double * __restrict__ d_volume_scale,
+    const unsigned long long * __restrict__ d_init_u,
+    const unsigned long long * __restrict__ d_final_u,
+    double * __restrict__ d_box,
+    const double * __restrict__ d_box_output,
+    double * __restrict__ d_x,
+    const double * __restrict__ d_x_output,
+    int * __restrict__ num_accepted,
+    int * __restrict__ num_attempted
+) {
+    const int idx = blockIdx.x*blockDim.x + threadIdx.x;
+    if (idx >= N) {
+        return;
+    }
 
-    step_++;
-    if (step_ % interval_ != 0) {
+    const double volume = d_box[0*3+0] * d_box[1*3+1] * d_box[2*3+2];
+    const double new_volume = volume+d_volume_delta[0];
+    const double energy_delta = FIXED_TO_FLOAT<double>(d_final_u[0] - d_init_u[0]);
+    const double w = energy_delta + pressure*d_volume_delta[0] - num_molecules * kt * std::log(new_volume/volume);
+
+    const bool rejected = w > 0 && rand[1] > std::exp(-w/kt);
+    if(idx == 0) {
+        if (!rejected) {
+            num_accepted[0]++;
+        }
+        num_attempted[0]++;
+        if(num_attempted[0] >= 10) {
+            if (num_accepted[0] < 0.25*num_attempted[0]) {
+                d_volume_scale[0] /= 1.1;
+                // Reset the counters
+                num_attempted[0] = 0;
+                num_accepted[0] = 0;
+            } else if (num_accepted[0] > 0.75*num_attempted[0]) {
+                d_volume_scale[0] = min(d_volume_scale[0]*1.1, volume*0.3);
+                // Reset the counters
+                num_attempted[0] = 0;
+                num_accepted[0] = 0;
+            }
+        }
+    }
+    if (rejected) {
         return;
     }
-    std::vector<double> h_box(9);
-    gpuErrchk(cudaMemcpy(&h_box[0], d_box, 9*sizeof(d_box), cudaMemcpyDeviceToHost));
+    // If the mc move was accepted copy all of the data into place
 
-    double volume = h_box[0*3+0]*h_box[1*3+1]*h_box[2*3+2];
+    if (idx < 9) {
+        d_box[idx] = d_box_output[idx];
+    }
 
-    if(volume_scale_ == 0) {
-        volume_scale_ = 0.01*volume;
+    #pragma unroll
+    for (int i = 0; i < 3; i ++) {
+        d_x[idx*3+i] = d_x_output[idx*3+i];
     }
+}
 
-    const int num_molecules = group_idxs_.size();
 
-    const int tpb = 32;
-    const int blocks = (num_grouped_atoms_ + tpb - 1) / tpb;
 
-    // Compute the energy of the modified system.
+void MonteCarloBarostat::reset_counters(){
+    gpuErrchk(cudaMemset(d_num_accepted_, 0, sizeof(*d_num_accepted_)));
```

Comment:
Did you want this to be Async/running in a stream?

Reply:
Nope, setting the values directly in k_decide_move, this is just a utility function for when we set a new interval/pressure